instance/models/database_server.py

```diff
@@ -0,0 +1,232 @@
+                    database_server.password, password,
+                    database_server.port, port,
+                )
+
+    def select_random(self):
+        """
+        Select a database server for a new instance.
+
+        The current implementation selects one of the database servers that accept new clients at random.
+        If no database server accepts new clients, DatabaseServer.DoesNotExist is raised.
+        """
+        self._create_default()
+
+        # The set of servers might change between retrieving the server count and retrieving the random server,
+        # so we make this atomic.
+        with transaction.atomic():
+            servers = self.filter(accepts_new_clients=True)  # pylint: disable=no-member
+            count = servers.count()
+            if not count:
+                raise self.model.DoesNotExist(  # pylint: disable=no-member
+                    "No configured DatabaseServer accepts new clients."
+                )
+            return servers[random.randrange(count)]
+
+
+class DatabaseServer(ValidateModelMixin, TimeStampedModel):
+    """
+    DatabaseServer: Abstract parent class for database server models.
+    """
+    objects = DatabaseServerManager()
+
+    class Meta:
+        abstract = True
+
+    # Host name or IP address of this database server
+    hostname = models.CharField(max_length=128, unique=True)
+
+    # Port to use when accessing this database server
+    port = models.PositiveIntegerField(blank=True)
+
+    # Credentials for accessing this database server.
+    # User identified by these credentials must have necessary permissions
+    # to create databases, create users, grant privileges on this database server.
+    username = models.CharField(max_length=64, blank=True)
+    password = models.CharField(max_length=128, blank=True)
+
+    # Does this database server currently accept new clients (i.e., instances)?
+    accepts_new_clients = models.BooleanField(default=True)
+
+    @property
+    def protocol(self):
+        """
+        Protocol to use for accessing this database server.
+        """
+        raise NotImplementedError("Specific database server implementations must define this.")
+
+    @property
+    def url(self):
+        """
+        Full URL for this database server.
+        """
+        url = '{protocol}://'.format(protocol=self.protocol)
+        if self.username:
+            url += self.username
+            if self.password:
+                url += ':{password}'.format(password=self.password)
+            url += '@'
+        url += self.hostname
+        return url
+
+    def __str__(self):
+        return self.hostname
+
+    def settings_match(self, username, password, port):
+        """
+        Return True if settings of this database server match settings passed to this method, else False.
+        """
+        return self.username == username and self.password == password and self.port == port
```

Comment:
There's a lot of finagling due to the default port number (lines [187-188](https://github.com/open-craft/opencraft/pull/151/files#diff-9c5bd8a78df5d6dfd0289a644e03cc8eR187), [197-198](https://github.com/open-craft/opencraft/pull/151/files#diff-9c5bd8a78df5d6dfd0289a644e03cc8eR197), [220-221](https://github.com/open-craft/opencraft/pull/151/files#diff-9c5bd8a78df5d6dfd0289a644e03cc8eR220), [230-231](https://github.com/open-craft/opencraft/pull/151/files#diff-9c5bd8a78df5d6dfd0289a644e03cc8eR230)).

Could you instead have a `default_port` property that each `DatabaseServer` subclass must override?  Then, you could move `set_field_defaults()` into the base class, and update`DatabaseServer.settings_match()` to call it too, removing the need for the subclass overrides:

```
def settings_match(self, username, password, port):
    if port is None:
        port = self.default_port
    return self.username == username and self.password == password and self.port == port 
```